website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,764 @@
+ **/
+MetaSchema.prototype.flatQuestions = function() {
+    var self = this;
+
+    var flat = [];
+
+    $.each(self.schema.pages, function(i, page) {
+        $.each(page.questions, function(qid, question) {
+            flat.push(question);
+        });
+    });
+    return flat;
+};
+
+/**
+ * @class Draft
+ * Model for DraftRegistrations
+ *
+ * @param {Object} params
+ * @param {String} params.pk: primary key of Draft
+ * @param {Object} params.registration_schema: data to be passed to MetaSchema constructor
+ * @param {Object} params.registration_metadata: saved Draft data
+ * @param {User} params.initiator
+ * @param {Date} params.initiated
+ * @param {Boolean} params.is_pending_review
+ * @param {Boolean} params.is_approved
+ * @param {Date} params.updated
+ * @param {Object} params.urls
+ * @param {String} params.urls.edit
+ * @param {String} params.urls.before_register
+ * @param {String} params.urls.register
+ * @property {Float} completion: percent completion of schema
+ **/
+var Draft = function(params, metaSchema) {
+    var self = this;
+
+    self.pk = params.pk;
+    self.metaSchema = metaSchema || new MetaSchema(params.registration_schema);
+    self.schema = ko.pureComputed(function() {
+        return self.metaSchema.schema;
+    });
+    self.schemaData = params.registration_metadata || {};
+
+    self.initiator = params.initiator;
+    self.initiated = new Date(params.initiated);
+    self.updated = new Date(params.updated);
+
+    self.urls = params.urls || {};
+
+
+    self.fulfills = params.fulfills || [];
+    self.isPendingReview = params.is_pending_review;
+    self.isApproved = params.is_approved;
+
+    self.requiresApproval = ko.pureComputed(function() {
+        return self.metaSchema && self.metaSchema.requiresApproval;
+    });
+    self.fulfills = ko.pureComputed(function() {
+        return self.metaSchema ? self.metaSchema.fulfills : [];
+    });
+
+    self.completion = ko.computed(function() {
+        var total = 0;
+        var complete = 0;
+        if (self.schemaData) {
+            var schema = self.schema();
+            $.each(schema.pages, function(i, page) {
+                $.each(page.questions, function(qid, question) {
+                    var q = self.schemaData[qid];
+                    if (q && (q.value || '').trim()) {
+                        complete++;
+                    }
+                    total++;
+                });
+            });
+            return Math.ceil(100 * (complete / total));
+        }
+        return 0;
+    });
+};
+Draft.prototype.preRegisterPrompts = function(response, confirm) {
+    var self = this;
+    bootbox.confirm({
+        size: 'large',
+        title: language.registerConfirm,
+        message: $osf.joinPrompts(response.prompts),
+        callback: function(result) {
+            if (result) {
+                confirm();
+            }
+        }
+    });
+};
+Draft.prototype.preRegisterErrors = function(response, confirm) {
```

Comment:
With the xhr changes below, this function and `preRegisterErrors` may not be necessary. (since as far as I can tell, they only exist to clean up success/error handling cases)